src/stavisky_lab_to_nwb/stavisky/staviskydecodinginterface.py

```diff
@@ -0,0 +1,166 @@
+            data=logits,
+            unit="n.a.",
+            timestamps=timestamps,
+            description="Log-probabilities of the 39 phonemes plus silence and space between words, as "
+            + "predicted by an RNN decoder",
+            comments="The 41 columns correspond to, in order, silence, the space between words, and "
+            + "the 39 phonemes of the CMU pronouncing dictionary in alphabetical order",
+        )
+
+        # add to processing module
+        processing_module.add_data_interface(logits_timeseries)
+
+        # close redis client
+        r.close()
+
+        return nwbfile
+
+
+class StaviskyDecodedTextInterface(BaseDataInterface):
+    """Decoded text interface for Stavisky Redis conversion"""
+
+    def __init__(
+        self,
+        port: int,
+        host: str,
+    ):
+        super().__init__(port=port, host=host)
+
+    def add_to_nwbfile(
+        self,
+        nwbfile: NWBFile,
+        metadata: dict,
+        # TODO: can also smooth timestamps for these if it turns out to be useful
+    ):
+        # initialize redis client and check connection
+        r = redis.Redis(
+            port=self.source_data["port"],
+            host=self.source_data["host"],
+        )
+        r.ping()
+
+        # get processing module
+        module_name = "decoding"
+        module_description = "Contains decoder outputs for real-time speech decoding"
+        processing_module = get_module(nwbfile=nwbfile, name=module_name, description=module_description)
+
+        # get reference time for timestamps
+        # session_start_time = metadata["NWBFile"]["session_start_time"].timestamp()
+        session_start_time = np.frombuffer(r.xrange("metadata")[0][1][b"startTime"], dtype=np.float64).item()
+
+        # loop through data and log decoder output
+        decoder_output = r.xrange("binned:decoderOutput:stream")
+        decoded_timestamps = []
+        decoded_text = []
+        last_text = ""
+        for entry in decoder_output:
+            if b"partial_decoded_sentence" in entry[1].keys():
+                curr_text = str(entry[1][b"partial_decoded_sentence"], "utf-8").strip()
+                if curr_text != last_text:  # if decoded text changes, log it
+                    timestamp = int(entry[0].split(b"-")[0]) / 1000.0 - session_start_time
+                    decoded_text.append(curr_text)
+                    decoded_timestamps.append(timestamp)
+                    last_text = curr_text
+            elif b"final_decoded_sentence" in entry[1].keys():  # reset on trial end
+                last_text = ""
+                # timestamp = int(entry[0].split(b'-')[0]) / 1000. - session_start_time
+                # # TODO: determine if this is necessary. does the model decide when to stop,
+                # # or is it externally controlled? if the former, then a STOP event might be useful
+                # decoded_text.append("STOP")
+                # decoded_timestamps.append(timestamp)
```

Comment:
Have we asked about this yet?

Reply:
Just wondering - we can leave it here in the meantime as long as we don't forget by the end

Reply:
ah whoops forgot to respond to this one. I believe the end of the trial is manually triggered with a button press, not by the model, which is why I commented out that code. But, I haven't asked them yet to make sure - didn't want to overwhelm them while they're working out timing issues. If a few more days pass without any updates from them, I will probably just ask anyway though so we're not sitting on this too long